File: 1450-delete-leaves-with-a-given-value/delete-leaves-with-a-given-value.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode removeLeafNodes(TreeNode root, int target) {
        return remove(root, target);
    }
    public TreeNode remove(TreeNode root, int target) {
        if(root == null) {
            return null;
        }
        if(root.left == null && root.right == null) {
            if(root.val == target) {
                return null;
            }
            return root;
        }
        TreeNode left = remove(root.left, target);
        TreeNode right = remove(root.right, target);
        root.left = left;
        root.right = right;
        if(left == null && right == null && root.val == target) {
            return null;
        }
        return root;
    }
}

File: 1450-delete-leaves-with-a-given-value/delete-leaves-with-a-given-value.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    TreeNode* removeLeaf(TreeNode* root, int target) {
        if(root == nullptr) {
            return nullptr;
        }
        root->left = removeLeaf(root->left, target);
        root->right = removeLeaf(root->right, target);
        if(root->left == nullptr && root->right == nullptr) {
            if(root->val == target) {
                root = nullptr;
            }
        }
        return root;
    }

    TreeNode* removeLeafNodes(TreeNode* root, int target) {
        return removeLeaf(root, target);
    }
};